File: python/revo1/revo1_touch_collector.py
"""
Revo1 Touch (Capacitive) - High-Performance Data Collector Example with Trajectory Control

This example demonstrates high-performance data collection for capacitive touch sensors
while simultaneously running trajectory control:
- Motor status data collection at 100Hz
- Capacitive touch sensor data collection at 100Hz
- Trajectory control at 100Hz
- Efficient data collection without GIL overhead
- Observe motor response and touch feedback in real-time

Hardware Support:
- ✅ Revo1 Touch

Usage:
    python revo1_touch_collector.py
"""

import asyncio
import sys
import time
from revo1_utils import *
from utils import setup_shutdown_event

# Trajectory control parameters
TRAJ_LEN = 20           # Number of trajectory points
CTRL_INTERVAL = 0.01    # Control interval: 10ms (100Hz)
ENABLE_CONTROL = True   # Enable/disable trajectory control


async def main():
    """Main function: High-performance capacitive touch data collection"""

    # Setup shutdown event listener
    shutdown_event = setup_shutdown_event(logger)

    # Open Modbus connection
    logger.info("Connecting to device...")
    (client, detected_slave_id) = await open_modbus_revo1()
    slave_id = detected_slave_id

    # Get device info
    device_info = await client.get_device_info(slave_id)
    logger.info(f"Device info: {device_info.description}")

    # Verify device has touch sensor
    if not device_info.is_revo1_touch():
        logger.error("This device does not support touch sensor!")
        sys.exit(1)

    logger.info("Detected capacitive touch sensor")

    # ========================================================================
    # Create shared buffers
    # ========================================================================

    logger.info("Creating shared buffers...")
    motor_buffer = libstark.MotorStatusBuffer(max_size=1000)
    touch_buffer = libstark.TouchStatusBuffer(max_size=1000)
    logger.info("Shared buffers created")

    # ========================================================================
    # Enable touch sensor
    # ========================================================================

    logger.info("Enabling touch sensor...")
    bits = 0x1F  # Enable all 5 fingers
    await client.touch_sensor_setup(slave_id, bits)
    await asyncio.sleep(1)

    bits = await client.get_touch_sensor_enabled(slave_id)
    logger.info(f"Touch sensor enabled status: {(bits & 0x1F):05b}")

    # ========================================================================
    # Create and start data collector - Capacitive mode
    # ========================================================================

    logger.info("Creating data collector (Capacitive mode)...")

    # Collection frequencies (platform-dependent)
    # Linux: 50Hz motor, 10Hz touch (Revo1 is slower than Revo2)
    # Windows/macOS: 20Hz motor, 10Hz touch (conservative)
    import platform
    is_linux = platform.system() == 'Linux'
    motor_frequency = 50 if is_linux else 20  # Revo1 motor is slower
    touch_frequency = 10  # Touch: 10Hz (same for all platforms)
    
    logger.info(f"Platform: {platform.system()}, Motor: {motor_frequency}Hz, Touch: {touch_frequency}Hz")

    collector = libstark.DataCollector.new_capacitive(
        ctx=client,
        motor_buffer=motor_buffer,
        touch_buffer=touch_buffer,
        slave_id=slave_id,
        motor_frequency=motor_frequency,
        touch_frequency=touch_frequency,
        enable_stats=True,
    )

    logger.info("Data collector created")

    # ========================================================================
    # Start data collection
    # ========================================================================

    logger.info("Starting data collector...")
    collector.start()
    logger.info("Data collector started (background thread)")
    logger.info("Please touch the sensor to see touch data...")

    # ========================================================================
    # Start trajectory control task (optional)
    # ========================================================================

    stop_flag = {'value': False}
    index_ref = {'value': 0}
    control_task = None

    if ENABLE_CONTROL:
        logger.info("Initializing trajectory control...")
        # Generate trajectory (0-100% range)
        # Note: Thumb trajectory will be auto-clamped to max 50% if needed
        trajectory = init_cosine_trajectory(traj_len=TRAJ_LEN)
        finger_id = libstark.FingerId.Ring  # Control ring finger

        control_task = asyncio.create_task(
            trajectory_control_task(
                client, slave_id, finger_id, trajectory,
                ctrl_interval=CTRL_INTERVAL,
                stop_flag_ref=stop_flag,
                index_ref=index_ref
            )
        )
        logger.info("Trajectory control task started")

    # ========================================================================
    # Start data processing tasks
    # ========================================================================

    # Start data processing tasks with intervals matching collection frequencies
    asyncio.create_task(process_motor_data(motor_buffer, index_ref if ENABLE_CONTROL else None))
    asyncio.create_task(process_touch_data(touch_buffer, touch_frequency))

    logger.info("Data processing tasks started")
    logger.info("Press Ctrl+C to stop...")

    # Wait for shutdown event
    await shutdown_event.wait()

    # ========================================================================
    # Stop and cleanup
    # ========================================================================

    logger.info("Stopping trajectory control...")
    if ENABLE_CONTROL and control_task:
        stop_flag['value'] = True
        await control_task
        logger.info("Trajectory control stopped")

    logger.info("Stopping data collector...")
    collector.stop()
    # Give a small delay to ensure logs are flushed
    await asyncio.sleep(0.1)
    logger.info("Data collector stopped")

    # Cleanup resources
    libstark.modbus_close(client)
    logger.info("Modbus client closed")
    sys.exit(0)


async def process_motor_data(motor_buffer, index_ref=None):
    """
    Process motor data

    Args:
        motor_buffer: Motor status buffer
        index_ref: Optional trajectory index reference for correlation
    """
    logger.info("Motor data processing task started")
    process_count = 0

    while True:
        try:
            if motor_buffer.is_empty():
                await asyncio.sleep(0.01)
                continue

            # Batch get data
            data_batch = motor_buffer.pop_all()

            if len(data_batch) > 0:
                if process_count % 2 == 0:  # Print every 2 times
                    traj_info = f" [Traj Index: {index_ref['value']}]" if index_ref else ""
                    logger.debug(f"[{process_count}]{traj_info} Processed {len(data_batch)} motor data")

                    # Print latest data
                    latest = data_batch[-1]
                    # logger.info(f"Positions: {latest.positions[:6]}")
                    # logger.info(f"Speeds: {latest.speeds[:6]}")
                    # logger.info(f"Currents: {latest.currents[:6]}")
                    logger.info(f"Finger Detail: {latest.description}")

                process_count += 1

            await asyncio.sleep(0.01)  # 10ms processing interval

        except Exception as e:
            logger.error(f"Motor data processing error: {e}")
            await asyncio.sleep(1)


async def process_touch_data(touch_buffer, frequency):
    """Process capacitive touch data
    
    Args:
        touch_buffer: Touch data buffer
        frequency: Collection frequency in Hz (determines processing interval)
    """
    # Calculate processing interval based on frequency
    # Use 2x frequency for processing to avoid buffer overflow
    process_interval = 1.0 / (frequency * 2)  # seconds
    
    logger.info(f"Capacitive touch data processing task started (interval: {process_interval*1000:.1f}ms)")
    process_count = 0

    # Finger names
    finger_names = ["Thumb", "Index", "Middle", "Ring", "Pinky"]

    while True:
        try:
            # Check buffer data count
            data_counts = touch_buffer.len_all()
            total_count = sum(data_counts)

            if total_count == 0:
                await asyncio.sleep(process_interval)
                continue

            # Get all finger touch data
            all_data = touch_buffer.pop_all()

            if process_count % 10 == 0:  # Print every 10 times
                logger.info(f"[{process_count}] Processed {len(all_data[0])} touch data")

            # Print each finger's latest data
            for finger_idx in range(5):
                if len(all_data[finger_idx]) > 0:
                    # Get latest data
                    latest_item: libstark.TouchFingerItem = all_data[finger_idx][-1]

                    # Check if there is contact
                    is_contact = latest_item.normal_force1 > 50  # Threshold: 50

                    if is_contact:
                        print(
                            f"  {finger_names[finger_idx]}: "
                            f"normal_force={latest_item.normal_force1}, "
                            f"tangential_force={latest_item.tangential_force1}, "
                            f"proximity={latest_item.self_proximity1} "
                            f"[CONTACT]"
                        )
                    else:
                        logger.debug(
                            f"  {finger_names[finger_idx]}: "
                            f"normal_force={latest_item.normal_force1}"
                        )

            process_count += 1
            await asyncio.sleep(process_interval)

        except Exception as e:
            logger.error(f"Touch data processing error: {e}")
            await asyncio.sleep(1)


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        logger.info("User interrupted")
        sys.exit(0)
    except Exception as e:
        logger.error(f"Error: {e}", exc_info=True)
        sys.exit(1)

File: linux/revo1/revo1_touch_collector.cpp
#include "stark-sdk.h"
#include "../common/stark_common.h"
#include "../common/trajectory_control.h"
#include <stdio.h>
#include <unistd.h>
#include <signal.h>

/**
 * Revo1 Touch (Capacitive) - High-Performance Data Collector Example
 * 
 * This example demonstrates high-performance data collection for capacitive touch sensors:
 * - Motor status data collection at 100Hz (Linux) or 20Hz (Windows/macOS)
 * - Capacitive touch sensor data collection at 100Hz (Linux) or 10Hz (Windows/macOS)
 * - Efficient data transfer via shared buffer API
 * - Lock-free buffer design for minimal overhead
 * - Optional trajectory control for testing motor response
 * - Suitable for real-time tactile feedback and control
 * 
 * Hardware Support:
 * - ✅ Revo1 Touch (Capacitive Sensor)
 * 
 * Configuration:
 *   #define ENABLE_CONTROL 1    // Enable trajectory control (0=disabled, 1=enabled)
 *   #define TRAJ_LEN 20         // Number of trajectory points
 *   #define CTRL_FREQUENCY 100  // Control frequency in Hz
 * 
 * Compile:
 *   make revo1_touch_collector
 * 
 * Run:
 *   ./revo1_touch_collector
 */

// ============================================================================
// Configuration
// ============================================================================

#define ENABLE_CONTROL 1      // Enable trajectory control (0=disabled, 1=enabled)
#define TRAJ_LEN 20           // Number of trajectory points
#define CTRL_FREQUENCY 100    // Control frequency in Hz

// Global variable for signal handling
static volatile int keep_running = 1;

void signal_handler(int signum) {
  printf("\n[INFO] Received signal %d, stopping...\n", signum);
  keep_running = 0;
}

int main(int argc, char const *argv[]) {
  // Setup signal handlers
  setup_signal_handlers();
  signal(SIGINT, signal_handler);
  signal(SIGTERM, signal_handler);

  // Initialize SDK
  init_cfg(STARK_PROTOCOL_TYPE_MODBUS, LOG_LEVEL_INFO);
  
  // Auto-detect device
  auto cfg = auto_detect_modbus_revo1("/dev/ttyUSB0", true);
  if (cfg == NULL) {
    fprintf(stderr, "[ERROR] Failed to auto-detect Modbus device.\n");
    return -1;
  }
  
  uint8_t slave_id = cfg->slave_id;
  printf("[INFO] Detected device on %s, slave_id=%d, baudrate=%d\n", 
         cfg->port_name, slave_id, cfg->baudrate);
  
  // Open device
  auto handle = modbus_open(cfg->port_name, cfg->baudrate);
  free_device_config(cfg);
  
  if (handle == NULL) {
    fprintf(stderr, "[ERROR] Failed to open device.\n");
    return -1;
  }

  // Get device info
  auto info = stark_get_device_info(handle, slave_id);
  if (info != NULL) {
    printf("[INFO] Device: SKU=%d, SN=%s, FW=%s\n",
           (uint8_t)info->sku_type, info->serial_number, info->firmware_version);
    
    // Check if device has touch sensor
    bool has_touch = (info->hardware_type == STARK_HARDWARE_TYPE_REVO1_TOUCH);
    
    if (!has_touch) {
      fprintf(stderr, "[ERROR] Device does not have touch sensor!\n");
      free_device_info(info);
      return -1;
    }
    
    printf("[INFO] Device has capacitive touch sensor\n");
    free_device_info(info);
  }

  // Enable touch sensor
  printf("[INFO] Enabling touch sensor...\n");
  stark_enable_touch_sensor(handle, slave_id, 0x1F);  // Enable all 5 fingers
  usleep(1000 * 1000); // Wait 1 second

  // ========================================================================
  // Create shared buffers
  // ========================================================================
  
  printf("[INFO] Creating shared buffers...\n");
  
  // Motor status buffer (max 1000 items)
  auto motor_buffer = motor_buffer_new(1000);
  if (motor_buffer == NULL) {
    fprintf(stderr, "[ERROR] Failed to create motor buffer.\n");
    return -1;
  }
  
  // Touch status buffer (max 1000 items per finger)
  auto touch_buffer = touch_buffer_new(1000);
  if (touch_buffer == NULL) {
    fprintf(stderr, "[ERROR] Failed to create touch buffer.\n");
    motor_buffer_free(motor_buffer);
    return -1;
  }

  // ========================================================================
  // Create and start data collector - Capacitive mode
  // ========================================================================
  
  printf("[INFO] Creating data collector (Capacitive mode)...\n");
  
  // Platform-dependent frequencies
  // Linux: 50Hz motor, 10Hz touch (Revo1 is slower)
  // Platform-specific frequencies
  #ifdef __linux__
    const uint32_t motor_frequency = 100;  // Linux: 100Hz (Revo1)
    const uint32_t touch_frequency = 100;  // Touch: 100Hz
  #else
    const uint32_t motor_frequency = 20;   // Windows/macOS: 20Hz (conservative)
    const uint32_t touch_frequency = 10;   // Touch: 10Hz
  #endif
  
  printf("[INFO] Motor frequency: %dHz, Touch frequency: %dHz\n", 
         motor_frequency, touch_frequency);
  
  auto collector = data_collector_new_capacitive(
    handle,                    // Device context
    motor_buffer,              // Motor buffer
    touch_buffer,              // Touch buffer
    slave_id,                  // Slave ID
    motor_frequency,           // Motor frequency (platform-dependent)
    touch_frequency,           // Touch frequency (platform-dependent)
    1                          // Enable stats
  );
  
  if (collector == NULL) {
    fprintf(stderr, "[ERROR] Failed to create data collector.\n");
    touch_buffer_free(touch_buffer);
    motor_buffer_free(motor_buffer);
    return -1;
  }

  // Start data collection
  printf("[INFO] Starting data collector...\n");
  if (data_collector_start(collector) != 0) {
    fprintf(stderr, "[ERROR] Failed to start data collector.\n");
    data_collector_free(collector);
    touch_buffer_free(touch_buffer);
    motor_buffer_free(motor_buffer);
    return -1;
  }

  printf("[INFO] Data collector started successfully!\n");
  printf("[INFO] Capacitive mode: Motor %dHz + Touch %dHz\n", motor_frequency, touch_frequency);
  printf("[INFO] Please touch the sensor to see touch data...\n");

  // ========================================================================
  // Optional: Start trajectory control
  // ========================================================================
  
  #if ENABLE_CONTROL
  printf("[INFO] Initializing trajectory control...\n");
  
  // Generate cosine trajectory (0-1000 range)
  // Note: Thumb trajectory will be auto-clamped to max 500 if needed
  uint16_t *trajectory = init_cosine_trajectory(TRAJ_LEN, 0, 1000);
  if (trajectory == NULL) {
    fprintf(stderr, "[ERROR] Failed to create trajectory.\n");
    data_collector_stop(collector);
    data_collector_free(collector);
    touch_buffer_free(touch_buffer);
    motor_buffer_free(motor_buffer);
    return -1;
  }
  
  // Initialize trajectory control for Ring finger
  TrajectoryControl traj_ctrl;
  trajectory_control_init(
    &traj_ctrl, handle, slave_id,
    STARK_FINGER_ID_RING,  // Control Ring finger
    trajectory, TRAJ_LEN, CTRL_FREQUENCY
  );
  
  // Start trajectory control
  if (trajectory_control_start(&traj_ctrl) != 0) {
    fprintf(stderr, "[ERROR] Failed to start trajectory control.\n");
    free(trajectory);
    data_collector_stop(collector);
    data_collector_free(collector);
    touch_buffer_free(touch_buffer);
    motor_buffer_free(motor_buffer);
    return -1;
  }
  
  printf("[INFO] Trajectory control started (Index finger, %dHz)\n", CTRL_FREQUENCY);
  #endif
  
  printf("[INFO] Press Ctrl+C to stop...\n\n");

  // ========================================================================
  // Main loop: Read data from buffers
  // ========================================================================
  
  // Allocate arrays for reading data
  const size_t MAX_MOTOR_DATA = 1000;
  const size_t MAX_TOUCH_DATA = 1000;
  
  MotorStatusData motor_data[MAX_MOTOR_DATA];
  TouchFingerItem touch_data[5][MAX_TOUCH_DATA];
  size_t motor_count = 0;
  size_t touch_counts[5] = {0};
  
  const char* finger_names[] = {"Thumb", "Index", "Middle", "Ring", "Pinky"};
  int loop_count = 0;
  
  while (keep_running) {
    usleep(100 * 1000); // Sleep 100ms
    loop_count++;
    
    // Read motor status every 10 loops (1 second)
    if (loop_count % 10 == 0) {
      size_t buffer_len = motor_buffer_len(motor_buffer);
      if (buffer_len > 0) {
        motor_buffer_pop_all(motor_buffer, motor_data, &motor_count);
        
        if (motor_count > 0) {
          auto latest = &motor_data[motor_count - 1];
          
          #if ENABLE_CONTROL
          size_t traj_idx = trajectory_control_get_index(&traj_ctrl);
          printf("[Motor] Count=%zu, Traj=%zu/%d, Latest: Pos=[%hu,%hu,%hu,%hu,%hu,%hu]\n",
                 motor_count, traj_idx, TRAJ_LEN,
                 latest->positions[0], latest->positions[1], latest->positions[2],
                 latest->positions[3], latest->positions[4], latest->positions[5]);
          #else
          printf("[Motor] Count=%zu, Latest: Pos=[%hu,%hu,%hu,%hu,%hu,%hu]\n",
                 motor_count,
                 latest->positions[0], latest->positions[1], latest->positions[2],
                 latest->positions[3], latest->positions[4], latest->positions[5]);
          #endif
        }
      }
    }
    
    // Read touch data
    bool has_touch_data = false;
    for (int finger = 0; finger < 5; finger++) {
      int result = touch_buffer_pop_finger(
        touch_buffer, finger, 
        touch_data[finger], &touch_counts[finger]
      );
      
      if (result == 0 && touch_counts[finger] > 0) {
        has_touch_data = true;
      }
    }
    
    if (has_touch_data) {
      printf("[Touch] Data counts: [%zu,%zu,%zu,%zu,%zu]\n",
             touch_counts[0], touch_counts[1], touch_counts[2],
             touch_counts[3], touch_counts[4]);
      
      // Print each finger's latest data
      for (int finger = 0; finger < 5; finger++) {
        if (touch_counts[finger] > 0) {
          auto latest_item = &touch_data[finger][touch_counts[finger] - 1];
          
          // Check if there is contact
          bool is_contact = latest_item->normal_force1 > 50;
          
          if (is_contact) {
            printf("  %s: normal_force=%d, tangential_force=%d, proximity=%d [CONTACT]\n",
                   finger_names[finger],
                   latest_item->normal_force1,
                   latest_item->tangential_force1,
                   latest_item->self_proximity1);
          }
        }
      }
      printf("\n");
    }
  }

  // ========================================================================
  // Cleanup
  // ========================================================================
  
  #if ENABLE_CONTROL
  printf("\n[INFO] Stopping trajectory control...\n");
  trajectory_control_stop(&traj_ctrl);
  free(trajectory);
  #endif
  
  printf("\n[INFO] Stopping data collector...\n");
  data_collector_stop(collector);
  
  printf("[INFO] Cleaning up...\n");
  data_collector_free(collector);
  touch_buffer_free(touch_buffer);
  motor_buffer_free(motor_buffer);
  
  printf("[INFO] Done!\n");
  
  return 0;
}
